File: project.R
# bike rental count prediction

#first clean R enviorment
rm(list=ls(all=T))

#set  working directory
setwd("C:/Users/himanshu gupta/Desktop/edwisor/project/2")
getwd()

#Load Libraries
x = c("ggplot2", "corrgram", "DMwR", "caret", "randomForest", "unbalanced", "C50", "dummies", "e1071", "Information",
      "MASS", "rpart", "gbm", "ROSE", "sampling", "DataCombine", "inTrees","gridExtra","scales","psych","gplots")

#install.packages(x)
lapply(x, require, character.only = TRUE)
rm(x)

#lets load the data
bike_rental_data = read.csv("day.csv")

#------------------------------------------------------#
#                 explore the data                     #

dim(bike_rental_data)
names(bike_rental_data)
head(bike_rental_data)
str(bike_rental_data)
summary(bike_rental_data)

#in our dataset some variables have no useful information for our prediction
#so it is better to remove those variables.it helps us to make useful inferences

#lets drop unnecessary variables
bike_rental_data = subset(bike_rental_data,select = -c(instant,dteday,casual,registered))

#------------------------------------------------------#
#                 data-preprocessing                   #

#missing value analysis
sapply(bike_rental_data, function(x) {
  sum(is.na(x))
})
# there are no missing values

#outlier analysis

cnames=c("temp",'atemp','windspeed','hum','cnt')
for (i in 1:length(cnames))
{
  assign(paste0("gn",i), ggplot(aes_string(y = (cnames[i]), x = "cnt"), data = subset(bike_rental_data))+ 
           stat_boxplot(geom = "errorbar", width = 0.5) +
           geom_boxplot(outlier.colour="green", fill = "grey" ,outlier.shape=18,
                        outlier.size=1, notch=FALSE) +
           theme(legend.position="bottom")+
           labs(y=cnames[i],x="count")+
           ggtitle(paste("Box plot of count for",cnames[i])))
}

#plotting boxplot
gridExtra::grid.arrange(gn1,gn2,ncol=2)
gridExtra::grid.arrange(gn3,gn4,ncol=2)
gridExtra::grid.arrange(gn5,ncol=1)

#lets remove outliers using boxplot
df = bike_rental_data
for(i in cnames){
  print(i)
  outliers = bike_rental_data[,i][bike_rental_data[,i] %in% boxplot.stats(bike_rental_data[,i])$out]
  print(length(outliers))
  bike_rental_data = bike_rental_data[which(!bike_rental_data[,i] %in% outliers),]
}

#lets plot boxplot after removing outliers
for (i in 1:length(cnames))
{
  assign(paste0("gn",i), ggplot(aes_string(y = (cnames[i]), x = "cnt"), data = subset(bike_rental_data))+ 
           stat_boxplot(geom = "errorbar", width = 0.5) +
           geom_boxplot(outlier.colour="green", fill = "grey" ,outlier.shape=18,
                        outlier.size=1, notch=FALSE) +
           theme(legend.position="bottom")+
           labs(y=cnames[i],x="cnt")+
           ggtitle(paste("Box plot of cnt for",cnames[i])))
}
#plotting Boxplot after removing outliers
gridExtra::grid.arrange(gn1,gn2,ncol=2)
gridExtra::grid.arrange(gn3,gn4,ncol=2)
gridExtra::grid.arrange(gn5,ncol=1)

#------------------------------------------------------#
#                 feature selection                    #

#find correlation matrix using corrplot and correlation plot using corrgram library
#FOR NUMERICAL VARIABLES

#save dataset after outlier analysis
df = bike_rental_data
#correlation matrix
cnames=c("temp","atemp","windspeed","hum")
sapply(bike_rental_data,class)
correlation_matrix = cor(bike_rental_data,cnames)
correlation_matrix
#correlation plot
corrgram(bike_rental_data[,cnames],order = F,upper.panel = panel.pie,
         text.panel = panel.txt,main = 'Correlation plot')

#From the correlation plot,we see that temp and atemp variables are correlated to each other
#so we need to remove atemp variable.

#perform annova test for categorical variables
catnames=c('season','yr','mnth','holiday','weekday','workingday','weathersit')
for (i in catnames) {
  print(i)
  anova = summary(aov(formula = cnt~bike_rental_data[,i],bike_rental_data))
  print(anova)
}
#based on the anova result, we can drop three variables named,
# holiday, weekday, workingday
#because these variables having the p-value > 0.05

#Dimension reduction
bike_rental_data = subset(bike_rental_data,select = -c(holiday,weekday,workingday,atemp))

#lets check after dimension reduction
dim(bike_rental_data)
head(bike_rental_data)

#------------------------------------------------------#
#                 feature scaling                      #

#check normality between the varaibles

cnames=c("temp","windspeed","hum","cnt")
for (i in cnames){
  print(i)
  normality = qqnorm(bike_rental_data[,i])
  
}

#already we plotted distrution between these variables,lets recall it
for(i in 1:length(cnames))
{
  assign(paste0("h",i),ggplot(aes_string(x=(cnames[i])),
                              data=subset(bike_rental_data))+
           geom_histogram(fill="blue",colour = "green")+geom_density()+
           scale_y_continuous(breaks =pretty_breaks(n=8))+
           scale_x_continuous(breaks = pretty_breaks(n=8))+
           theme_bw()+xlab(cnames[i])+ylab("Frequency")+
           ggtitle(paste("distribution plot for ",cnames[i])))
}
gridExtra::grid.arrange(h1,h2,h3,h4,ncol = 2)

#summary of the data
for (i in cnames) {
  print(i)
  print(summary(bike_rental_data[,i]))
  
}
#Based on the above inferences and plots,we can see that the variables are normalised.

# bivariate analysis for categorical variables
bivariate_categorical <-
  function(dataset, variable, targetVariable) {
    variable <- enquo(variable)
    targetVariable <- enquo(targetVariable)
    
    ggplot(
      data = dataset,
      mapping = aes_(
        x = rlang::quo_expr(variable),
        y = rlang::quo_expr(targetVariable),
        fill = rlang::quo_expr(variable)
      )
    ) +
      geom_boxplot() +
      theme(legend.position = "bottom") -> p
    plot(p)
    
  }

bivariate_continous <-
  function(dataset, variable, targetVariable) {
    variable <- enquo(variable)
    targetVariable <- enquo(targetVariable)
    ggplot(data = dataset,
           mapping = aes_(
             x = rlang::quo_expr(variable),
             y = rlang::quo_expr(targetVariable)
           )) +
      geom_point() +
      geom_smooth() -> q
    plot(q)
    
  }

bivariate_categorical(bike_rental_data, season, cnt)
bivariate_categorical(bike_rental_data, yr, cnt)
bivariate_categorical(bike_rental_data, mnth, cnt)
bivariate_categorical(bike_rental_data, weathersit, cnt)
bivariate_continous(bike_rental_data, temp, cnt)
bivariate_continous(bike_rental_data, hum, cnt)
bivariate_continous(bike_rental_data, windspeed, cnt)


#------------------------------------------------------#
#                 model devlopment                     #

#we can not pass categorical variables to regression problems
#so convert categorical variables into dummy variables
#saving our preprocessed data
df = bike_rental_data

#create dummies
library(dummies)
catnames = c('season','yr','mnth','weathersit')
bike_rental_data = dummy.data.frame(bike_rental_data,catnames)

#we have created dummies,lets check dimension and top 5 observations
dim(bike_rental_data)
head(bike_rental_data)

#divide the data into train and test
set.seed(1234)
train_index = sample(1:nrow(df), 0.8 * nrow(df))
train_data = bike_rental_data[train_index,]
test_data = bike_rental_data[-train_index,]

#------------------------------------------------------#
#                 (1) linear regression                #

#running regression model
lm_model = lm(cnt~. ,data = bike_rental_data)
#lets check performance of our modedl
summary(lm_model)
#Residual standard error: 787.3 on 696 degrees of freedom
#Multiple R-squared:  0.8388,	Adjusted R-squared:  0.8342 
#F-statistic: 181.1 on 20 and 696 DF,  p-value: < 2.2e-16

# Function for Error metrics to calculate the performance of model
#lets build function for MAPE
#calculate MAPE
MAPE = function(y, y1){
  mean(abs((y - y1)/y))
}

# Function for r2 to calculate the goodness of fit of model
rsquare=function(y,y1){
  cor(y,y1)^2
}

# Function for RMSE value 
RMSE = function(y,y1){
  difference = y - y1
  root_mean_square = sqrt(mean(difference^2))
}

#lets predict for train and test data
Predictions_LR_train = predict(lm_model,train_data[,-25])
Predictions_LR_test = predict(lm_model,test_data[,-25])

#let us check performance of our model

#mape calculation
LR_train_mape = MAPE(Predictions_LR_train,train_data[,25])
LR_test_mape = MAPE(test_data[,25],Predictions_LR_test)

#Rsquare calculation
LR_train_r2 = rsquare(train_data[,25],Predictions_LR_train)
LR_test_r2 = rsquare(test_data[,25],Predictions_LR_test)

#rmse calculation
LR_train_rmse = RMSE(train_data[,25],Predictions_LR_train)
LR_test_rmse = RMSE(test_data[,25],Predictions_LR_test)

print(LR_train_mape) #0.15
print(LR_test_mape) #0.18
print(LR_train_r2) #0.831
print(LR_test_r2) #0.867
print(LR_train_rmse) #789.6
print(LR_test_rmse) #717.2


#------------------------------------------------------#
#                 (2) decision tree                #

library(rpart)
DT_model = rpart(cnt ~ ., data = train_data, method = "anova")
DT_model


#predicting for train and test data
predictions_DT_train= predict(DT_model,train_data[,-25])
predictions_DT_test= predict(DT_model,test_data[,-25])

# MAPE calculation
DT_train_mape = MAPE(train_data[,25],predictions_DT_train)
DT_test_mape = MAPE(test_data[,25],predictions_DT_test)

# Rsquare calculation
DT_train_r2= rsquare(train_data[,25],predictions_DT_train)
DT_test_r2 = rsquare(test_data[,25],predictions_DT_test)

# RMSE calculation
DT_train_rmse = RMSE(train_data[,25],predictions_DT_train)
DT_test_rmse = RMSE(test_data[,25],predictions_DT_test)

print(DT_train_mape) #0.522
print(DT_test_mape) #0.243
print(DT_train_r2) #0.811
print(DT_test_r2) #0.798
print(DT_train_rmse) #833.848
print(DT_test_rmse) #885.59

#------------------------------------------------------#
#                 (3) Random Forest                #

#building random forest model
RF_model = randomForest(cnt~.,data = train_data,n.trees = 600)
print(RF_model)

#lets predict for both train and test data
predictions_RF_train = predict(RF_model,train_data[-25])
predictions_RF_test = predict(RF_model,test_data[-25])

#MAPE calculation
RF_train_mape = MAPE(predictions_RF_train,train_data[,25])
RF_test_mape = MAPE(predictions_RF_test,test_data[,25])

#Rsquare calculation
RF_train_r2 = rsquare(predictions_RF_train,train_data[,25])
RF_test_r2 = rsquare(predictions_RF_test,test_data[,25])

#RMSE calculation
RF_train_rmse = RMSE(train_data[,25],predictions_RF_train)
RF_test_rmse = RMSE(test_data[,25],predictions_RF_test)

print(RF_train_mape) #0.07
print(RF_test_mape) #0.12
print(RF_train_r2) #0.965
print(RF_test_r2) #0.910
print(RF_train_rmse) #371.18
print(RF_test_rmse) #593.84

#------------------------------------------------------#
#                   model selection                    #

Model_name = c('Linear regression',
               'Decision tree',
               'Random forest')

MAPE_train = c(LR_train_mape,DT_train_mape,
               RF_train_mape)

MAPE_test = c(LR_test_mape,DT_test_mape,
              RF_test_mape)

Rsquare_train = c(LR_train_r2,DT_train_r2,
                  RF_train_r2)

Rsquare_test = c(LR_test_r2,DT_test_r2,
                 RF_test_r2)

RMSE_train =  c(LR_train_rmse,DT_train_rmse,
                RF_train_rmse)

RMSE_test = c(LR_test_rmse,DT_test_rmse,
              RF_test_rmse)

FINAL_RESULTS = data.frame(Model_name,MAPE_train,MAPE_test,Rsquare_train,Rsquare_test,
                           RMSE_train,RMSE_test)

print(FINAL_RESULTS)


#Index  Model_name           MAPE_train  MAPE_test    Rsquare_train  Rsquare_test  RMSE_train  RMSE_test
#1      Linear regression    0.15497164  0.1829289    0.8311816      0.8671739     789.6785    717.2833
#2      Decision tree        0.52210598  0.2438791    0.8119266      0.7986807     833.4855    885.5906
#3      Random forest        0.07256787  0.1224177    0.9652738      0.9103488     371.1827    593.8403

# Based on the above inferences,we came to know that Random forest performs very well in our dataset
#so we are finalising that model.

File: project.py

# coding: utf-8

# In[612]:

#import working libraries
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sn
get_ipython().magic('matplotlib inline')
import statsmodels.api as sm

from scipy.stats import chi2_contingency
from statsmodels.formula.api import ols
from sklearn.linear_model import LinearRegression
from sklearn.cross_validation import train_test_split
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn import metrics


# In[613]:

#set working directory
os.chdir(r"C:\Users\himanshu gupta\Desktop\edwisor\project\2")


# In[614]:

#load data
bike_rental_data= pd.read_csv("day.csv")


# ### data exploration

# In[615]:

#dimension of data
bike_rental_data.shape


# In[616]:

#checking first 5 rows
bike_rental_data.head(5)


# In[617]:

#checking data type of all variables
bike_rental_data.dtypes


# In[618]:

# checking summary of the dataset
bike_rental_data.describe()


# In[619]:

#converting some useful variables into categorical variabels
categorical_var= ['season','yr','mnth','holiday','weekday','workingday','weathersit']
for a in categorical_var:
    bike_rental_data[a]=bike_rental_data[a].astype("category")


# In[620]:

#checking datatypes again
bike_rental_data.dtypes


# we will not use instant,dateday, casual and registered variable because they are not caryying useful information.

# ### data preprocessing

# ##### target variable distribution

# In[621]:

fig,(ax1,ax2) = plt.subplots(ncols=2)
fig.set_size_inches(10,6)
sn.distplot(bike_rental_data["cnt"],ax=ax1)
stats.probplot(bike_rental_data["cnt"], dist='norm', fit=True, plot=ax2)


# we can cleary see that cnt is very close to normal distribution.

# ##### missing value analysis

# In[622]:

bike_rental_data.isnull().sum()


# there are no missing values.

# ##### outliner analysis

# from above boxplot following things are clear:

# In[623]:

fig, axes = plt.subplots(nrows=2,ncols=2)
fig.set_size_inches(14,14)
sn.boxplot(data=bike_rental_data,y="cnt",orient='v',ax=axes[0][0])
sn.boxplot(data=bike_rental_data,y="cnt",x="season",orient='v',ax=axes[0][1])
sn.boxplot(data=bike_rental_data,y="cnt",x="weekday",orient="v",ax=axes[1][0])
sn.boxplot(data=bike_rental_data,y="cnt",x="workingday",orient="v",ax=axes[1][1])
axes[0][0].set(ylabel='cnt',title = "Boxplot of cnt")
axes[0][1].set(xlabel="season",ylabel="cnt",title="Boxplot for cnt vs season")
axes[1][0].set(xlabel="weekday", ylabel="cnt",title="Boxplot for cnt vs weekday")
axes[1][1].set(xlabel="workingday",ylabel="cnt",title="Boxplot for cnt vs workingday")


# (1)there are no outliers in count.
# (2)demands for bike is very low in spring season.

# from above boxplot following things are clear:
# (1)there are no outliers in count.
# (2)demands for bike is very low in spring season.

# In[624]:

fig, axes = plt.subplots(nrows=2,ncols=2)
fig.set_size_inches(14,14)
sn.boxplot(data=bike_rental_data,y="cnt",x="yr",orient='v',ax=axes[0][0])
sn.boxplot(data=bike_rental_data,y="cnt",x="mnth",orient='v',ax=axes[0][1])
sn.boxplot(data=bike_rental_data,y="cnt",x="holiday",orient='v',ax=axes[1][0])
sn.boxplot(data=bike_rental_data,y="cnt",x="weathersit",orient='v',ax=axes[1][1])
axes[0][0].set(xlabel="yr",ylabel="cnt",title="Boxplot for cnt vs yr")
axes[0][1].set(xlabel="mnth",ylabel="cnt",title="Boxplot for cnt vs mnth")
axes[1][0].set(xlabel="holiday",ylabel="cnt",title="Boxplot for cnt vs holiday")
axes[1][1].set(xlabel="weathersit",ylabel="cnt",title="Boxplot for cnt vs weathersit")


# from above boxplot following things are clear:

# (1) demands for bike is high in year 2011.
# (2) demands for bike is gardually incaresing from january to september and then started to decreasing.
# (3) demands for bike is high when there is clear weather.

# In[625]:

fig, axes = plt.subplots(nrows=2,ncols=2)
fig.set_size_inches(14,14)
sn.boxplot(data=bike_rental_data,y="temp",orient='v',ax=axes[0][0])
sn.boxplot(data=bike_rental_data,y="atemp",orient='v',ax=axes[0][1])
sn.boxplot(data=bike_rental_data,y="hum",orient='v',ax=axes[1][0])
sn.boxplot(data=bike_rental_data,y="windspeed",orient='v',ax=axes[1][1])
axes[0][0].set(ylabel="temp",title="Boxplot for temp")
axes[0][1].set(ylabel="atemp",title="Boxplot for atemp")
axes[1][0].set(ylabel="hum",title="Boxplot for hum")
axes[1][1].set(ylabel="windspeed",title="Boxplot for windspeed")


# From the above boxplot we can cleary see that there are:
# (1) outliers in windspeed.
# (2) inliers in humidity.

# In[626]:

#removal of outliers and inliers
cnames=["hum","windspeed"]
for i in cnames:
    print(i)
    q75, q25 = np.percentile(bike_rental_data.loc[:,i], [75 ,25])
    iqr = q75 - q25
    min = q25 - (iqr*1.5)
    max = q75 + (iqr*1.5)
    print(min)
    print(max)
    bike_rental_data = bike_rental_data.drop(bike_rental_data[bike_rental_data.loc[:,i] < min].index)
    bike_rental_data = bike_rental_data.drop(bike_rental_data[bike_rental_data.loc[:,i] > max].index)
    min = bike_rental_data.loc[bike_rental_data[i] < min,i] 
    max = bike_rental_data.loc[bike_rental_data[i] > max,i]


# subsituted inliers with minimum values and outliers with maximum values.

# In[627]:

#checking humidity and windspeed after removal of inliers and outliers
fig.set_size_inches(14,14)
sn.boxplot(data=bike_rental_data,y="hum").set_title("Boxplot of humidity")


# In[628]:

fig.set_size_inches(14,14)
sn.boxplot(data=bike_rental_data,y="windspeed").set_title("Boxplot of windspeed")


# ##### feature selection

# In[629]:

##Correlation analysis
#numeric variables
cnames=["temp","atemp","hum","windspeed","cnt"]
#Correlation plot
bike_rental_corr = bike_rental_data.loc[:,cnames]


# In[630]:

#Set the width and hieght of the plot
f, ax = plt.subplots(figsize=(9, 7))

#Generate correlation matrix
corr = bike_rental_corr.corr()

#Plot using seaborn library
sn.heatmap(corr, mask=np.zeros_like(corr, dtype=np.bool), cmap=sn.diverging_palette(220, 10, as_cmap=True),
            square=True, ax=ax)


# from the above plot,we came to know that both temp and atemp variables are carrying almost same information
# hence there is no need to continue with both variables.so we need to drop any one of the variables
# here we are dropping atemp variable.

# In[631]:

#Anova test for categorical variables(target variable is numeric)
#Save categorical variables
cat_names = ["season", "yr", "mnth", "holiday", "weekday", "workingday", "weathersit"]


# In[632]:

for i in cat_names:
    results = ols('cnt' + '~' + i, data = bike_rental_data).fit()
    aov_table = sm.stats.anova_lm(results, typ = 2)
    print(aov_table)


# based on the anova result, we are going to drop three variables holiday,weekday,workingday
# because these variables have the p-value > 0.05

# In[633]:

# Removing the variables which have p-value > 0.05 and are correlated variable or does not contain useful information and store into a new dataset
df = bike_rental_data.drop(['atemp', 'holiday','weekday','workingday','instant','dteday','casual','registered'], axis=1)
bike_rental_data=df.copy()


# In[634]:

#now check dimension of data
bike_rental_data.shape


# In[635]:

bike_rental_data


# ##### feature scaling

# In[636]:

col=["temp","hum","windspeed","cnt"]


# In[637]:

for i in col:
    print(i)
    sn.distplot(bike_rental_data[i],bins='auto',color='black')
    plt.title("distribution plot for "+i)
    plt.ylabel("density")
    plt.show()


# based on distribution plot we can clearly see that all the numeric variables are normalized.

# ##### bivariate analysis

# In[638]:

# Bivariate analysis of cnt and continous variables

fig,(ax1,ax2,ax3) = plt.subplots(ncols=3)
fig.set_size_inches(12,8)

sn.regplot(x="temp",y="cnt",data=bike_rental_data,ax=ax1)
sn.regplot(x="hum",y="cnt",data=bike_rental_data,ax=ax2)
sn.regplot(x="windspeed",y="cnt",data=bike_rental_data,ax=ax3)


# from above boxplot it is clear that bike count has:
# (1) positive linear relationship with temperature.
# (2) slightly negative linear relationship with humidity.
# (3) negative linear reltionship with windspeed.

# # model devlopment

# In[639]:

#In Regression problems, we can't directly pass categorical variables.so we need to convert all categorical variables 
#into dummy variables.
ccol=['season','yr','mnth','weathersit']

#  Converting categorical variables to dummy variables
df = pd.get_dummies(bike_rental_data,columns=ccol)
bike_rental_data=df


# In[640]:

#Divide the data into train and test set 

x= bike_rental_data.drop(['cnt'],axis=1)
y= bike_rental_data['cnt']
x_train,x_test,y_train,y_test= train_test_split(x,y,test_size=.25)


# In[641]:

# Function for Error metrics to calculate the performance of model
def MAPE(y_true,y_prediction):
    mape= np.mean(np.abs(y_true-y_prediction)/y_true)*100
    return mape


# ### linear regression model

# In[642]:

LinearRegression_model= sm.OLS(y_train,x_train).fit()
print(LinearRegression_model.summary())


# In[643]:

# Model prediction on  train data
LinearRegression_train= LinearRegression_model.predict(x_train)

# Model prediction on test data
LinearRegression_test= LinearRegression_model.predict(x_test)

# Model performance on train data
MAPE_train= MAPE(y_train,LinearRegression_train)

# Model performance on test data
MAPE_test= MAPE(y_test,LinearRegression_test)

# r2 value for train data
r2_train= r2_score(y_train,LinearRegression_train)

# r2 value for test data-
r2_test=r2_score(y_test,LinearRegression_test)

# RMSE value for train data
RMSE_train = np.sqrt(metrics.mean_squared_error(y_train,LinearRegression_train))

# RMSE value for test data
RMSE_test = np.sqrt(metrics.mean_squared_error(y_test,LinearRegression_test))

print("Mean Absolute % Error for train data="+str(MAPE_train))
print("Mean Absolute % error for test data="+str(MAPE_test))
print("R^2_score for train data="+str(r2_train))
print("R^2_score for test data="+str(r2_test))
print("RMSE for train data="+str (RMSE_train))
print("RMSE for test data="+str(RMSE_test))


# In[644]:

Error_MetricsLT = {'Model Name': ['Linear Regression'],
                 'MAPE_Train':[MAPE_train],
                 'MAPE_Test':[MAPE_test],
                 'R-squared_Train':[r2_train],
                 'R-squared_Test':[r2_test],
                 'RMSE_train':[RMSE_train],
                 'RMSE_test':[RMSE_test]}

LinearRegression_Results = pd.DataFrame(Error_MetricsLT)


# In[645]:

LinearRegression_Results


# ### random forest

# In[646]:

# Random Forest for regression
RF_model= RandomForestRegressor(n_estimators=80).fit(x_train,y_train)

# Prediction on train data
RF_train= RF_model.predict(x_train)

# Prediction on test data
RF_test= RF_model.predict(x_test)

# MAPE For train data
MAPE_train= MAPE(y_train,RF_train)

# MAPE For test data
MAPE_test= MAPE(y_test,RF_test)

# Rsquare  For train data
r2_train= r2_score(y_train,RF_train)

# Rsquare  For test data
r2_test=r2_score(y_test,RF_test)

# RMSE value for train data
RMSE_train = np.sqrt(metrics.mean_squared_error(y_train,RF_train))

# RMSE value for test data
RMSE_test = np.sqrt(metrics.mean_squared_error(y_test,RF_test))

print("Mean Absolute % Error for train data="+str(MAPE_train))
print("Mean Absolute % Error for test data="+str(MAPE_test))
print("R^2_score for train data="+str(r2_train))
print("R^2_score for test data="+str(r2_test))
print("RMSE for train data="+str (RMSE_train))
print("RMSE for test data="+str(RMSE_test))


# In[647]:

Error_MetricsRF = {'Model Name': ['Random Forest'],
                 'MAPE_Train':[MAPE_train],
                 'MAPE_Test':[MAPE_test],
                 'R-squared_Train':[r2_train],
                 'R-squared_Test':[r2_test],
                 'RMSE_train':[RMSE_train],
                 'RMSE_test':[RMSE_test]}
                   
RandomForest_Results = pd.DataFrame(Error_MetricsRF)
RandomForest_Results


# ### decision tree

# In[648]:

# Decision tree for regression
DecisionTree_model= DecisionTreeRegressor(max_depth=3).fit(x_train,y_train)

# Model prediction on train data
DecisionTree_train= DecisionTree_model.predict(x_train)

# Model prediction on test data
DecisionTree_test= DecisionTree_model.predict(x_test)

# Model performance on train data
MAPE_train= MAPE(y_train,DecisionTree_train)

# Model performance on test data
MAPE_test= MAPE(y_test,DecisionTree_test)

# r2 value for train data
r2_train= r2_score(y_train,DecisionTree_train)

# r2 value for test data
r2_test=r2_score(y_test,DecisionTree_test)

# RMSE value for train data
RMSE_train = np.sqrt(metrics.mean_squared_error(y_train,DecisionTree_train))

# RMSE value for test data
RMSE_test = np.sqrt(metrics.mean_squared_error(y_test,DecisionTree_test))

print("Mean Absolute Precentage Error for train data="+str(MAPE_train))
print("Mean Absolute Precentage Error for test data="+str(MAPE_test))
print("R^2_score for train data="+str(r2_train))
print("R^2_score for test data="+str(r2_test))
print("RMSE for train data="+str(RMSE_train))
print("RMSE for test data="+str(RMSE_test))


# In[649]:

Error_MetricsDT = {'Model Name': ['Decision Tree'],
                 'MAPE_Train':[MAPE_train],
                 'MAPE_Test':[MAPE_test],
                 'R-squared_Train':[r2_train],
                 'R-squared_Test':[r2_test],
                 'RMSE_train':[RMSE_train],
                 'RMSE_test':[RMSE_test]}
                   
DecisionTree_Results = pd.DataFrame(Error_MetricsDT)
DecisionTree_Results 


# From above results Random Forest & linear regression both model have optimum values and this algorithms are good for our data 

# In[650]:

#saveing the out put of finalized model (random forest)

input = y_test.reset_index()
predicted = pd.DataFrame(RF_test,columns = ['predicted'])
Final_result = predicted.join(input)
Final_result


# In[651]:

Final_result.to_csv("Final_results_python.csv",index=False)


# In[ ]:



